src/openassetio-core/src/hostApi/ManagerFactory.cpp

```diff
@@ -109,6 +115,21 @@ ManagerPtr ManagerFactory::defaultManagerForInterface(
 
   const std::string_view identifier = config["manager"]["identifier"].value_or("");
 
+  // Function to substitute $config_dir or ${config_dir} with the
+  // absolute, canonicalised directory of the TOML config file.
+  const auto substituteConfigDir =
+      [configDir = std::filesystem::canonical(configPath).parent_path().string()](
+          std::string& str) -> std::string& {
```

Comment:
[fix] As discussed offline, this by-ref return is actually a pessimisation, causing an additional copy.

I'd love to link the quick-bench benchmark i did on this, but as usual the online servers timeout on anything significant.
Therefore i'll put the test code below, as well as a screenshot of the result from my local instance.
<details>
  <summary>Test Comparisons</summary>
  
```cpp
#include <string>
#include <unordered_map>
#include <variant>

std::string& substituteConfigDirRef(std::string& str)
{
    return str;
}

std::string substituteConfigDirVal(std::string str)
{
  return str;
}


static void StringRefReturn(benchmark::State& state) {
  std::unordered_map<int, std::variant<bool, int, float, std::string>> infoDict;
  int counter = 0;
  std::string setting = "a string that isn't sso optimised";
  for (auto _ : state) {
    infoDict.insert({counter, substituteConfigDirRef(setting)});
    counter++;
    benchmark::DoNotOptimize(infoDict);
    benchmark::DoNotOptimize(substituteConfigDirRef);
  }
}
// Register the function as a benchmark
BENCHMARK(StringRefReturn);

static void StringRefReturnWithMove(benchmark::State& state) {
  std::unordered_map<int, std::variant<bool, int, float, std::string>> infoDict;
  int counter = 0;
  std::string setting = "a string that isn't sso optimised";
  for (auto _ : state) {
    infoDict.insert({counter, std::move(substituteConfigDirRef(setting))});
    counter++;
    benchmark::DoNotOptimize(infoDict);
    benchmark::DoNotOptimize(substituteConfigDirRef);
  }
}
// Register the function as a benchmark
BENCHMARK(StringRefReturnWithMove);

static void StringValueReturn(benchmark::State& state) {
  std::unordered_map<int, std::variant<bool, int, float, std::string>> infoDict;
  int counter = 0;
  std::string setting = "a string that isn't sso optimised";
  for (auto _ : state) {
    infoDict.insert({counter, substituteConfigDirVal(std::move(setting))});
    counter++;
    benchmark::DoNotOptimize(infoDict);
    benchmark::DoNotOptimize(substituteConfigDirVal);
  }
}
BENCHMARK(StringValueReturn);
```

</details>

![image](https://github.com/OpenAssetIO/OpenAssetIO/assets/114349211/cab4c470-32c2-4c52-8897-73aae6b9ee3b)

Reply:
Great find, tweaked in e465f308

Reply:
:+1: 

(That .get() to r-value is super weird and uncomfortable, at a glance you'd think you were forgetting to move.)